### PR TITLE
Resource Command Support

### PR DESCRIPTION
**Pull request by tlmii**
First pieces of the resource command support:

- Forked the FluentMenuButton component to make an AspireMenuButton component that had a few extra features (different item binding/commanding structure, ability to change the appearance of the button, ability to have only an icon of our choosing for the button). Already chatting with @vnbaaij about what to do long term
- Temporarily implemented commands in the local resource server. I won't merge until that is removed, but it'll make it easier for folks to test it out. Start/Stop commands always fail. Restart always succeeds. Start/Restart have confirmation message, Stop does not. Some resources (randomly) have no commands at all.
- If no commands are present on a particular resource, the button is still there, but its disabled
- If no commands are present on any resources, the column doesn't appear at all
- Toast component pops up for Success/Failure of the command. Failure has a View Logs link to take you to the console logs for the resource. Not 100% positive that's useful, but wanted to show it for discussion
- MessageBox component used for confirmation

Here's a quick recording of it in action:

ResourceCommandsDark

Some notes:
1. I had to hack around a glitch where the menu would appear offscreen very briefly, causing the horizontal scrollbar to appear. That's what the extra CSS for the grid does. This is only an issue when its the last column (or otherwise near the right edge of the screen)
2. I initially had no text in the column header for the commands. I didn't think it was necessary. But there's an odd rendering issue with the grid where the resize handle doesn't appear if there's no text. So I just made it say Commands
3. I have some special casing in the error handling for when ExecuteResourceCommand isn't implemented by the resource server. I'm not sure that's really necessary, but it seemed appropriate at this state where we are just starting to implement that (and indeed it is not implemented in the local server once I remove my change)
4. As mentioned above, I'll remove the changes to Hosting's DashboardService.cs and Partials.cs before this gets merged.
5. The toasts disappear after 7s (the default). We should consider whether that should be increased/decreased. We should also make sure it plays well with screen readers (if it doesn't, that should probably be fixed in the component long term).
6. We should consider if there needs to be keyboard shortcuts for the toasts.
7. I did not implement merging commands that exist on resource types into the ones that exist on a resource as part of this PR. At the moment, we don't really use the resource type object from the API in a way that made that straightforward. Since the first use case for these commands doesn't use it either, I think we can do that as a second step.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2572)

### CODE REVIEW

File: src/Aspire.Dashboard/Model/ResourceViewModel.cs
Comment (kvenkatrajan) [inline src/Aspire.Dashboard/Model/ResourceViewModel.cs:65]: Assuming we would want to localize the displayName?

Comment (tlmii) [reply src/Aspire.Dashboard/Model/ResourceViewModel.cs:65]: I assume we will want that localized somehow, along with things like custom statuses and whatnot. But I'm not sure what the plan is for that. Will discuss with folks, but I think that can come in a separate PR.

Comment (adamint) [reply src/Aspire.Dashboard/Model/ResourceViewModel.cs:65]: We should consider sending user culture to hosting and letting hosting localize

**Comment by davidfowl** [comment]
Very cool @tlmii. I'm planning to look at implementing restart in Aspire.Hosting. The change looks like I would expect! We should think about future operations that may need more than a confirmation and make sure the resource server contract can expand to it.

**Comment by tlmii** [comment]
> Very cool @tlmii. I'm planning to look at implementing restart in Aspire.Hosting. The change looks like I would expect! 

Nice - let me know what you think of the experience when you start that - I'm expecting that'll generate more feedback since it'll widen the audience for this UX.

> We should think about future operations that may need more than a confirmation and make sure the resource server contract can expand to it.

Yeah I think this area will need more attention as we get a better idea of the commands that need to run through it. Even now with the confirmation dialog, it feels a bit forced. For example, what should the buttons say? Yes/No? OK/Cancel? Whatever we pick forces the language that can be used in the confirmation dialog. So even that could be more flexible - though we probably don't want to go the route of implementing an entire MessageBox API in the resource server api.

**Comment by tlmii** [comment]
I went ahead and removed the test code. If anyone wants to run it, they can grab the 2nd to last commit (https://github.com/dotnet/aspire/pull/2572/commits/30e7655ef3b920123a41d3a3c43a72ccd4ac93e8 as of this writing) to make testing easier - or just use the code in the original commit (https://github.com/dotnet/aspire/pull/2572/commits/c1a126274e32cf82fd7cf2a86930bcedc3ec4aab#diff-53cce2ea4b16776f87f7628686480c626c9fce25ea5a28eb46dd95f1dc72717f and https://github.com/dotnet/aspire/pull/2572/commits/c1a126274e32cf82fd7cf2a86930bcedc3ec4aab#diff-d35f6c3b9acebacf4a61b2055f987e92d1a207f889ac9892a6c4ad92f6492186) as a template for your own.

**Comment by JamesNK** [comment]
I ran TestShop on the branch and didn't see the column. I'm guessing that means there are no commands.

How do we test this?

**Comment by davidfowl** [comment]
Lets get this merged. I need to design the app model API for exposing this 😄

**Review by kvenkatrajan — APPROVED** [review]

File: src/Aspire.Dashboard/Components/Controls/AspireMenuButton.razor
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/AspireMenuButton.razor:20]: nit: Indenting if 5 spaces instead of 4.

File: src/Aspire.Dashboard/Components/Pages/Resources.razor
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Pages/Resources.razor:89]: There were issues with the metrics table and dynamic columns like this. Basically, the table data would change but it would take a moment before the columns updated. I don't think that's an issue here, but something to keep an eye on.

File: src/Aspire.Dashboard/Components/Pages/Resources.razor.css
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Pages/Resources.razor.css:37]: Add comments for what these do

File: src/Aspire.Dashboard/Model/DashboardClient.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Model/DashboardClient.cs:391]: Combine passed in cancellation token with the one from the CTS on the class.

File: src/Aspire.Dashboard/Model/DashboardClient.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Model/DashboardClient.cs:399]: Get text from resources

File: src/Aspire.Dashboard/Components/Controls/AspireMenuButton.razor.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/AspireMenuButton.razor.cs:15]: Should this be a field?

**Review by JamesNK — APPROVED** [review]
Feedback:

* The `...` button disable state is very subtle. It needs to be more obvious. Instead of changing the opacity, maybe the color should be grey.
* The menu goes off the bottom of the screen:
image
  Fixed an issue like this in metrics. Does the control you're using have the same setting? https://github.com/dotnet/aspire/pull/2301

Comments + issues here should be fixed but I trust you to do it in a follow up. Can merge now to unblock people

**Comment by tlmii** [comment]
Thanks @JamesNK. I fixed a few of the things real quick before merge. Will tackle the rest in the morning in a separate PR.

File: src/Aspire.Dashboard/Model/DashboardClient.cs
Comment (JamesNK) [inline src/Aspire.Dashboard/Model/DashboardClient.cs:391]:
```suggestion
            using var combinedTokens = CancellationTokenSource.CreateLinkedTokenSource(_clientCancellationToken, cancellationToken);
```

**Comment by JamesNK** [comment]
Need to rebase. `_clientCancellationToken` is from a change yesterday.

**Comment by tlmii** [comment]
@JamesNK

> The ... button disable state is very subtle. It needs to be more obvious. Instead of changing the opacity, maybe the color should be grey.

Yeash, I'm not real fond of it but it's how the built-in control appears to work. I wanted to look at that specifically to see if there was something built-in to handle it or if we needed to go custom with it.

**Comment by tlmii** [comment]
> * The menu goes off the bottom of the screen:
>   image
>   Fixed an issue like this in metrics. Does the control you're using have the same setting? [Update Blazor Fluent UI package to 4.4.2 preview #2301](https://github.com/dotnet/aspire/pull/2301)

Filed https://github.com/dotnet/aspire/issues/2661 to track https://github.com/microsoft/fluentui-blazor/issues/1640 to address this.

File: src/Aspire.Dashboard/Model/DashboardClient.cs
Comment (tlmii) [reply src/Aspire.Dashboard/Model/DashboardClient.cs:403]: After chatting about this offline, we're going to leave this hard-coded for the moment. Other error messages coming across the wire won't be localized either (at least for now) and rather than address the potential problems here (bringing loc into DashboardClient or changing the viewmodel to handle what was already just special-casing of the not implemented exception) we can address it later if it becomes a problem and perhaps those issues will be different/handled/non-existent by then.

**Comment by tlmii** [comment]
> @JamesNK
> 
> > The ... button disable state is very subtle. It needs to be more obvious. Instead of changing the opacity, maybe the color should be grey.
> 
> Yeah, I'm not real fond of it but it's how the built-in control appears to work. I wanted to look at that specifically to see if there was something built-in to handle it or if we needed to go custom with it.

Figured it out - PR here: https://github.com/dotnet/aspire/pull/2670

**Pull request merged**
